File: src/domain/models/tool_adapter.py
"""Tool adapter protocol and concrete adapters for the Tool Translation context.

Each adapter translates a DomainModel into tool-native configuration files.
Adapters are pure domain logic -- they produce content strings without I/O.

Quality gate display and file globs are read from a StackProfile so that
output is stack-aware. GenericProfile produces configs with no quality gates.

Supported tools:
- Claude Code: .claude/CLAUDE.md
- Cursor: AGENTS.md + .cursor/rules/project-conventions.mdc
- Roo Code: AGENTS.md + .roomodes + .roo/rules/project-conventions.md
- OpenCode: AGENTS.md + .opencode/rules/project-conventions.md + opencode.json
"""

from __future__ import annotations

import json
from typing import TYPE_CHECKING, Protocol, runtime_checkable

from src.domain.models.tool_config import ConfigSection

if TYPE_CHECKING:
    from src.domain.models.domain_model import DomainModel
    from src.domain.models.stack_profile import StackProfile


@runtime_checkable
class ToolAdapterProtocol(Protocol):
    """Interface for translating a DomainModel into tool-native config sections."""

    def translate(self, model: DomainModel, profile: StackProfile) -> tuple[ConfigSection, ...]:
        """Translate a DomainModel into config sections for a specific tool.

        Args:
            model: A finalized DomainModel with bounded contexts and UL.
            profile: Stack profile providing quality gate display and file globs.

        Returns:
            Tuple of ConfigSection value objects ready for output.
        """
        ...


# ---------------------------------------------------------------------------
# Shared constants
# ---------------------------------------------------------------------------

_AFTER_CLOSE_PROTOCOL_TEXT = """## After-Close Protocol

After every `bd close <id>`, run these steps:

1. **Ripple review** -- `bin/bd-ripple <id> "<what this ticket produced>"`
2. **Review flagged tickets** -- `bd query label=review_needed`, read ripple comments,
   draft updates, present to user for approval.
   For **dependent tickets**: run a compatibility check -- read the delivered source files
   and trace the interfaces the flagged ticket assumes. Cite file:line for every claim.
3. **Follow-up tickets** -- create using beads templates, set dependencies
4. **Groom next ticket** -- `bd ready`, run grooming checklist on top pick
"""

# ---------------------------------------------------------------------------
# Shared content builders
# ---------------------------------------------------------------------------


def _build_ubiquitous_language_section(model: DomainModel) -> str:
    """Render the ubiquitous language glossary as markdown."""
    lines: list[str] = ["## Ubiquitous Language", ""]
    lines.append("| Term | Definition | Context |")
    lines.append("|------|-----------|---------|")
    lines.extend(
        f"| {entry.term} | {entry.definition} | {entry.context_name} |"
        for entry in model.ubiquitous_language.terms
    )
    lines.append("")
    return "\n".join(lines)


def _build_bounded_context_section(model: DomainModel) -> str:
    """Render bounded contexts and their classifications as markdown."""
    lines: list[str] = ["## Bounded Contexts", ""]
    for ctx in model.bounded_contexts:
        classification = ctx.classification.value if ctx.classification else "unclassified"
        lines.append(f"- **{ctx.name}** ({classification}): {ctx.responsibility}")
    lines.append("")
    return "\n".join(lines)


def _build_ddd_layer_rules() -> str:
    """Render DDD layer dependency rules as markdown."""
    return """## DDD Layer Rules

- `domain/` has ZERO external dependencies (no frameworks, no DB, no HTTP)
- `application/` depends on `domain/` and `ports/` (interfaces only)
- `infrastructure/` implements `ports/` and depends on external libraries
- Dependencies flow inward: infrastructure -> application -> domain
"""


def _build_after_close_section() -> str:
    """Render the after-close protocol section for CLAUDE.md / AGENTS.md.

    Reuses the same protocol text as MEMORY.md for consistency.
    """
    return _AFTER_CLOSE_PROTOCOL_TEXT


def _build_quality_gates(profile: StackProfile) -> str:
    """Render quality gate commands from the stack profile.

    Returns profile.quality_gate_display (code block format) for stacks
    that have quality gates, or empty string for GenericProfile.
    """
    return profile.quality_gate_display


def _build_memory_md(model: DomainModel, profile: StackProfile) -> str:
    """Build MEMORY.md encoding the DDD agile work process.

    Content is kept under 200 lines (Claude Code's critical window).
    Quality gates are included only when profile provides them.
    """
    parts: list[str] = [
        "# Project Memory",
        "",
        _build_memory_beads_workflow(),
        _build_memory_after_close_protocol(),
        _build_memory_grooming_checklist(),
        _build_memory_bounded_contexts(model),
        _build_memory_ubiquitous_language(model),
    ]
    gates = _build_quality_gates(profile)
    if gates:
        parts.append(gates)
    return "\n".join(parts)


def _build_memory_beads_workflow() -> str:
    """Render beads workflow commands for MEMORY.md."""
    return """## Beads Workflow

```bash
bd ready                         # Find available work
bd show <id>                     # View ticket details
bd update <id> --status in_progress  # Claim a ticket
bd close <id>                    # Close completed ticket
bin/bd-ripple <id> "<summary>"   # Flag dependents (ripple review)
bd query label=review_needed     # See tickets needing review
bd label remove <id> review_needed   # Clear flag after review
```
"""


def _build_memory_after_close_protocol() -> str:
    """Render the after-close protocol for MEMORY.md."""
    return _AFTER_CLOSE_PROTOCOL_TEXT


def _build_memory_grooming_checklist() -> str:
    """Render the grooming checklist for MEMORY.md."""
    return """## Grooming Checklist

Before claiming a ticket:

1. **Template compliance** -- description follows beads template
2. **Freshness check** -- `bd label list <id>` for `review_needed`
3. **PRD traceability** -- `/prd-traceability <id>` to verify capability coverage
4. **DDD alignment** -- bounded context boundaries respected
5. **Ubiquitous language** -- terms match DDD.md glossary
6. **TDD & SOLID** -- RED/GREEN/REFACTOR phases documented
7. **Acceptance criteria** -- testable checkboxes, edge cases, coverage >= 80%
"""


def _build_memory_bounded_contexts(model: DomainModel) -> str:
    """Render bounded contexts summary for MEMORY.md."""
    lines: list[str] = ["## Bounded Contexts", ""]
    for ctx in model.bounded_contexts:
        classification = ctx.classification.value if ctx.classification else "unclassified"
        lines.append(f"- **{ctx.name}** ({classification}): {ctx.responsibility}")
    lines.append("")
    return "\n".join(lines)


def _build_memory_ubiquitous_language(model: DomainModel) -> str:
    """Render ubiquitous language summary for MEMORY.md."""
    lines: list[str] = ["## Ubiquitous Language", ""]
    lines.append("| Term | Definition | Context |")
    lines.append("|------|-----------|---------|")
    lines.extend(
        f"| {entry.term} | {entry.definition} | {entry.context_name} |"
        for entry in model.ubiquitous_language.terms
    )
    lines.append("")
    return "\n".join(lines)


def _build_agents_md(model: DomainModel, profile: StackProfile) -> str:
    """Build a generic AGENTS.md with project conventions from the domain model."""
    parts: list[str] = [
        "# Project Conventions",
        "",
        _build_ubiquitous_language_section(model),
        _build_bounded_context_section(model),
        _build_ddd_layer_rules(),
        _build_after_close_section(),
    ]
    gates = _build_quality_gates(profile)
    if gates:
        parts.append(gates)
    return "\n".join(parts)


# ---------------------------------------------------------------------------
# Concrete adapters
# ---------------------------------------------------------------------------


class ClaudeCodeAdapter:
    """Generates .claude/CLAUDE.md and .claude/memory/MEMORY.md for Claude Code."""

    def translate(self, model: DomainModel, profile: StackProfile) -> tuple[ConfigSection, ...]:
        """Translate DomainModel into Claude Code configuration."""
        parts: list[str] = [
            "# CLAUDE.md",
            "",
            _build_ubiquitous_language_section(model),
            _build_bounded_context_section(model),
            _build_ddd_layer_rules(),
            _build_after_close_section(),
        ]
        gates = _build_quality_gates(profile)
        if gates:
            parts.append(gates)
        content = "\n".join(parts)
        memory_content = _build_memory_md(model, profile)
        return (
            ConfigSection(
                file_path=".claude/CLAUDE.md",
                content=content,
                section_name="Claude Code config",
            ),
            ConfigSection(
                file_path=".claude/memory/MEMORY.md",
                content=memory_content,
                section_name="Claude Code memory",
            ),
        )


class CursorAdapter:
    """Generates AGENTS.md and .cursor/rules/project-conventions.mdc for Cursor."""

    def translate(self, model: DomainModel, profile: StackProfile) -> tuple[ConfigSection, ...]:
        """Translate DomainModel into Cursor configuration."""
        agents_content = _build_agents_md(model, profile)

        mdc_parts: list[str] = [
            "---",
            "description: Project conventions derived from domain model",
            f"globs: {profile.file_glob}",
            "---",
            "",
            _build_ddd_layer_rules(),
        ]
        gates = _build_quality_gates(profile)
        if gates:
            mdc_parts.append(gates)
        mdc_content = "\n".join(mdc_parts)

        return (
            ConfigSection(
                file_path="AGENTS.md",
                content=agents_content,
                section_name="Cursor agents",
            ),
            ConfigSection(
                file_path=".cursor/rules/project-conventions.mdc",
                content=mdc_content,
                section_name="Cursor rules",
            ),
        )


class RooCodeAdapter:
    """Generates AGENTS.md, .roomodes, and .roo/rules/project-conventions.md."""

    def translate(self, model: DomainModel, profile: StackProfile) -> tuple[ConfigSection, ...]:
        """Translate DomainModel into Roo Code configuration."""
        agents_content = _build_agents_md(model, profile)

        roomodes = json.dumps(
            {
                "customModes": [
                    {
                        "slug": "ddd-developer",
                        "name": "DDD Developer",
                        "description": "Follows domain-driven design conventions",
                    }
                ]
            },
            indent=2,
        )

        rules_parts: list[str] = [
            "# Project Conventions",
            "",
            _build_ddd_layer_rules(),
        ]
        gates = _build_quality_gates(profile)
        if gates:
            rules_parts.append(gates)
        rules_content = "\n".join(rules_parts)

        return (
            ConfigSection(
                file_path="AGENTS.md",
                content=agents_content,
                section_name="Roo Code agents",
            ),
            ConfigSection(
                file_path=".roomodes",
                content=roomodes,
                section_name="Roo Code modes",
            ),
            ConfigSection(
                file_path=".roo/rules/project-conventions.md",
                content=rules_content,
                section_name="Roo Code rules",
            ),
        )


class OpenCodeAdapter:
    """Generates AGENTS.md, .opencode/rules/project-conventions.md, and opencode.json."""

    def translate(self, model: DomainModel, profile: StackProfile) -> tuple[ConfigSection, ...]:
        """Translate DomainModel into OpenCode configuration."""
        agents_content = _build_agents_md(model, profile)

        rules_parts: list[str] = [
            "# Project Conventions",
            "",
            _build_ddd_layer_rules(),
        ]
        gates = _build_quality_gates(profile)
        if gates:
            rules_parts.append(gates)
        rules_content = "\n".join(rules_parts)

        opencode_json = json.dumps(
            {
                "context": {
                    "include": ["AGENTS.md", ".opencode/rules/*.md"],
                }
            },
            indent=2,
        )

        return (
            ConfigSection(
                file_path="AGENTS.md",
                content=agents_content,
                section_name="OpenCode agents",
            ),
            ConfigSection(
                file_path=".opencode/rules/project-conventions.md",
                content=rules_content,
                section_name="OpenCode rules",
            ),
            ConfigSection(
                file_path="opencode.json",
                content=opencode_json,
                section_name="OpenCode config",
            ),
        )

File: internal/tooltranslation/domain/tool_adapter.go
package domain

import (
	"encoding/json"
	"fmt"
	"strings"

	"github.com/alto-cli/alto/internal/shared/domain/ddd"
	vo "github.com/alto-cli/alto/internal/shared/domain/valueobjects"
)

// ---------------------------------------------------------------------------
// Shared constants
// ---------------------------------------------------------------------------

const afterCloseProtocolText = `## After-Close Protocol

After every ` + "`bd close <id>`" + `, run these steps:

1. **Ripple review** -- ` + "`bin/bd-ripple <id> \"<what this ticket produced>\"`" + `
2. **Review flagged tickets** -- ` + "`bd query label=review_needed`" + `, read ripple comments,
   draft updates, present to user for approval
3. **Follow-up tickets** -- create using beads templates, set dependencies
4. **Groom next ticket** -- ` + "`bd ready`" + `, run grooming checklist on top pick
`

// agentPersonaTemplate is the YAML frontmatter + body template for Claude Code agents.
const agentPersonaTemplate = `---
name: %s
description: >
  %s
tools: %s
model: opus
permissionMode: %s
memory: project
---

%s
`

type agentPersonaDef struct {
	slug        string
	description string
	tools       string
	permission  string
	bodyBuilder func(model *ddd.DomainModel, profile vo.StackProfile) string
}

var defaultAgentPersonas = []agentPersonaDef{
	{
		slug:        "developer",
		description: "Implementation-focused developer agent for writing code, fixing bugs, and implementing features following Red/Green/Refactor.",
		tools:       "Read, Edit, Write, Grep, Glob, Bash",
		permission:  "acceptEdits",
		bodyBuilder: buildDeveloperBody,
	},
	{
		slug:        "tech-lead",
		description: "Technical lead for architecture review, DDD/SOLID compliance, and code quality gate enforcement.",
		tools:       "Read, Grep, Glob, Bash, Write, Edit",
		permission:  "default",
		bodyBuilder: buildTechLeadBody,
	},
	{
		slug:        "qa-engineer",
		description: "QA engineer for writing tests, validating coverage, and investigating failures.",
		tools:       "Read, Edit, Write, Grep, Glob, Bash",
		permission:  "acceptEdits",
		bodyBuilder: buildQAEngineerBody,
	},
	{
		slug:        "researcher",
		description: "Research agent for spike tickets, ADRs, and library evaluation.",
		tools:       "Read, Write, Edit, Grep, Glob, Bash, WebSearch, WebFetch",
		permission:  "default",
		bodyBuilder: buildResearcherBody,
	},
}

// ---------------------------------------------------------------------------
// Shared content builders
// ---------------------------------------------------------------------------

func buildUbiquitousLanguageSection(model *ddd.DomainModel) string {
	var b strings.Builder
	b.WriteString("## Ubiquitous Language\n\n")
	b.WriteString("| Term | Definition | Context |\n")
	b.WriteString("|------|-----------|---------|")
	for _, entry := range model.UbiquitousLanguage().Terms() {
		fmt.Fprintf(&b, "\n| %s | %s | %s |", entry.Term(), entry.Definition(), entry.ContextName())
	}
	b.WriteString("\n")
	return b.String()
}

func buildBoundedContextSection(model *ddd.DomainModel) string {
	var b strings.Builder
	b.WriteString("## Bounded Contexts\n\n")
	for _, ctx := range model.BoundedContexts() {
		classification := "unclassified"
		if ctx.Classification() != nil {
			classification = string(*ctx.Classification())
		}
		fmt.Fprintf(&b, "- **%s** (%s): %s\n", ctx.Name(), classification, ctx.Responsibility())
	}
	return b.String()
}

func buildDDDLayerRules() string {
	return `## DDD Layer Rules

- ` + "`domain/`" + ` has ZERO external dependencies (no frameworks, no DB, no HTTP)
- ` + "`application/`" + ` depends on ` + "`domain/`" + ` and ` + "`ports/`" + ` (interfaces only)
- ` + "`infrastructure/`" + ` implements ` + "`ports/`" + ` and depends on external libraries
- Dependencies flow inward: infrastructure -> application -> domain
`
}

func buildQualityGates(profile vo.StackProfile) string {
	return profile.QualityGateDisplay()
}

func buildAgentsMD(model *ddd.DomainModel, profile vo.StackProfile) string {
	var b strings.Builder
	b.WriteString("# Project Conventions\n\n")
	b.WriteString(buildUbiquitousLanguageSection(model))
	b.WriteString("\n")
	b.WriteString(buildBoundedContextSection(model))
	b.WriteString("\n")
	b.WriteString(buildDDDLayerRules())
	b.WriteString("\n")
	b.WriteString(afterCloseProtocolText)
	gates := buildQualityGates(profile)
	if gates != "" {
		b.WriteString("\n")
		b.WriteString(gates)
	}
	return b.String()
}

func buildMemoryMD(model *ddd.DomainModel, profile vo.StackProfile) string {
	var b strings.Builder
	b.WriteString("# Project Memory\n\n")
	b.WriteString(buildMemoryBeadsWorkflow())
	b.WriteString("\n")
	b.WriteString(afterCloseProtocolText)
	b.WriteString("\n")
	b.WriteString(buildMemoryGroomingChecklist())
	b.WriteString("\n")
	b.WriteString(buildMemoryBoundedContexts(model))
	b.WriteString("\n")
	b.WriteString(buildMemoryUbiquitousLanguage(model))
	gates := buildQualityGates(profile)
	if gates != "" {
		b.WriteString("\n")
		b.WriteString(gates)
	}
	return b.String()
}

func buildMemoryBeadsWorkflow() string {
	return "## Beads Workflow\n\n```bash\n" +
		"bd ready                         # Find available work\n" +
		"bd show <id>                     # View ticket details\n" +
		"bd update <id> --status in_progress  # Claim a ticket\n" +
		"bd close <id>                    # Close completed ticket\n" +
		"bin/bd-ripple <id> \"<summary>\"   # Flag dependents (ripple review)\n" +
		"bd query label=review_needed     # See tickets needing review\n" +
		"bd label remove <id> review_needed   # Clear flag after review\n" +
		"```\n"
}

func buildMemoryGroomingChecklist() string {
	return `## Grooming Checklist

Before claiming a ticket:

1. **Template compliance** -- description follows beads template
2. **Freshness check** -- ` + "`bd label list <id>`" + ` for ` + "`review_needed`" + `
3. **PRD traceability** -- ` + "`/prd-traceability <id>`" + ` to verify capability coverage
4. **DDD alignment** -- bounded context boundaries respected
5. **Ubiquitous language** -- terms match DDD.md glossary
6. **TDD & SOLID** -- RED/GREEN/REFACTOR phases documented
7. **Acceptance criteria** -- testable checkboxes, edge cases, coverage >= 80%
`
}

func buildMemoryBoundedContexts(model *ddd.DomainModel) string {
	var b strings.Builder
	b.WriteString("## Bounded Contexts\n\n")
	for _, ctx := range model.BoundedContexts() {
		classification := "unclassified"
		if ctx.Classification() != nil {
			classification = string(*ctx.Classification())
		}
		fmt.Fprintf(&b, "- **%s** (%s): %s\n", ctx.Name(), classification, ctx.Responsibility())
	}
	return b.String()
}

func buildMemoryUbiquitousLanguage(model *ddd.DomainModel) string {
	var b strings.Builder
	b.WriteString("## Ubiquitous Language\n\n")
	b.WriteString("| Term | Definition | Context |\n")
	b.WriteString("|------|-----------|---------|")
	for _, entry := range model.UbiquitousLanguage().Terms() {
		fmt.Fprintf(&b, "\n| %s | %s | %s |", entry.Term(), entry.Definition(), entry.ContextName())
	}
	b.WriteString("\n")
	return b.String()
}

// ---------------------------------------------------------------------------
// Agent persona body builders
// ---------------------------------------------------------------------------

func buildDeveloperBody(model *ddd.DomainModel, profile vo.StackProfile) string {
	var b strings.Builder
	b.WriteString("You are a **Developer** on this project.\n\n")
	b.WriteString("## Primary Responsibilities\n\n")
	b.WriteString("1. Implement features and fix bugs via beads tickets\n")
	b.WriteString("2. Follow Red/Green/Refactor strictly\n")
	b.WriteString("3. Respect bounded context boundaries\n\n")
	b.WriteString(buildBoundedContextSection(model))
	b.WriteString("\n")
	b.WriteString(buildUbiquitousLanguageSection(model))
	b.WriteString("\n")
	b.WriteString(buildDDDLayerRules())
	gates := buildQualityGates(profile)
	if gates != "" {
		b.WriteString("\n")
		b.WriteString(gates)
	}
	return b.String()
}

func buildTechLeadBody(model *ddd.DomainModel, profile vo.StackProfile) string {
	var b strings.Builder
	b.WriteString("You are a **Tech Lead** on this project.\n\n")
	b.WriteString("## Primary Responsibilities\n\n")
	b.WriteString("1. Review architecture for DDD/SOLID compliance\n")
	b.WriteString("2. Enforce bounded context boundaries\n")
	b.WriteString("3. Run quality gates before approving changes\n\n")
	b.WriteString(buildBoundedContextSection(model))
	b.WriteString("\n")
	b.WriteString(buildUbiquitousLanguageSection(model))
	b.WriteString("\n")
	b.WriteString(buildDDDLayerRules())
	gates := buildQualityGates(profile)
	if gates != "" {
		b.WriteString("\n")
		b.WriteString(gates)
	}
	return b.String()
}

func buildQAEngineerBody(model *ddd.DomainModel, profile vo.StackProfile) string {
	var b strings.Builder
	b.WriteString("You are a **QA Engineer** on this project.\n\n")
	b.WriteString("## Primary Responsibilities\n\n")
	b.WriteString("1. Write and run tests, validate coverage\n")
	b.WriteString("2. Investigate failures and produce root cause analysis\n")
	b.WriteString("3. Verify edge cases from multiple angles\n\n")
	b.WriteString(buildBoundedContextSection(model))
	b.WriteString("\n")
	b.WriteString(buildUbiquitousLanguageSection(model))
	b.WriteString("\n")
	b.WriteString(buildDDDLayerRules())
	gates := buildQualityGates(profile)
	if gates != "" {
		b.WriteString("\n")
		b.WriteString(gates)
	}
	return b.String()
}

func buildResearcherBody(model *ddd.DomainModel, profile vo.StackProfile) string {
	var b strings.Builder
	b.WriteString("You are a **Researcher** on this project.\n\n")
	b.WriteString("## Primary Responsibilities\n\n")
	b.WriteString("1. Evaluate libraries and tools for spike tickets\n")
	b.WriteString("2. Write ADRs and research reports\n")
	b.WriteString("3. Provide concrete facts before architectural decisions\n\n")
	b.WriteString(buildBoundedContextSection(model))
	b.WriteString("\n")
	b.WriteString(buildUbiquitousLanguageSection(model))
	return b.String()
}

func buildAgentPersonas(model *ddd.DomainModel, profile vo.StackProfile) []ConfigSection {
	var sections []ConfigSection
	for _, def := range defaultAgentPersonas {
		body := def.bodyBuilder(model, profile)
		content := fmt.Sprintf(agentPersonaTemplate,
			def.slug,
			def.description,
			def.tools,
			def.permission,
			body,
		)
		path := fmt.Sprintf(".claude/agents/%s.md", def.slug)
		sections = append(sections, NewConfigSection(path, content, "Agent persona"))
	}
	return sections
}

// ---------------------------------------------------------------------------
// ClaudeCodeAdapter
// ---------------------------------------------------------------------------

// ClaudeCodeAdapter generates .claude/CLAUDE.md and .claude/memory/MEMORY.md.
type ClaudeCodeAdapter struct{}

// NewClaudeCodeAdapter creates a ClaudeCodeAdapter.
func NewClaudeCodeAdapter() *ClaudeCodeAdapter { return &ClaudeCodeAdapter{} }

// Translate implements ToolAdapter.
func (a *ClaudeCodeAdapter) Translate(model *ddd.DomainModel, profile vo.StackProfile) []ConfigSection {
	var b strings.Builder
	b.WriteString("# CLAUDE.md\n\n")
	b.WriteString(buildUbiquitousLanguageSection(model))
	b.WriteString("\n")
	b.WriteString(buildBoundedContextSection(model))
	b.WriteString("\n")
	b.WriteString(buildDDDLayerRules())
	b.WriteString("\n")
	b.WriteString(afterCloseProtocolText)
	gates := buildQualityGates(profile)
	if gates != "" {
		b.WriteString("\n")
		b.WriteString(gates)
	}

	memoryContent := buildMemoryMD(model, profile)

	base := []ConfigSection{
		NewConfigSection(".claude/CLAUDE.md", b.String(), "Claude Code config"),
		NewConfigSection(".claude/memory/MEMORY.md", memoryContent, "Claude Code memory"),
	}
	return append(base, buildAgentPersonas(model, profile)...)
}

// ---------------------------------------------------------------------------
// CursorAdapter
// ---------------------------------------------------------------------------

func buildCursorDomainExpertRule(model *ddd.DomainModel) string {
	var b strings.Builder
	b.WriteString("---\n")
	b.WriteString("description: Domain terminology and bounded context reference\n")
	b.WriteString("globs: **/*\n")
	b.WriteString("alwaysApply: true\n")
	b.WriteString("---\n\n")
	b.WriteString("# Domain Expert Reference\n\n")
	b.WriteString("Use these terms consistently throughout the codebase.\n\n")
	b.WriteString(buildUbiquitousLanguageSection(model))
	b.WriteString("\n")
	b.WriteString(buildBoundedContextSection(model))
	return b.String()
}

// CursorAdapter generates AGENTS.md and .cursor/rules/project-conventions.mdc.
type CursorAdapter struct{}

// NewCursorAdapter creates a CursorAdapter.
func NewCursorAdapter() *CursorAdapter { return &CursorAdapter{} }

// Translate implements ToolAdapter.
func (a *CursorAdapter) Translate(model *ddd.DomainModel, profile vo.StackProfile) []ConfigSection {
	agentsContent := buildAgentsMD(model, profile)

	var mdc strings.Builder
	fmt.Fprintf(&mdc, "---\ndescription: Project conventions derived from domain model\nglobs: %s\n---\n\n", profile.FileGlob())
	mdc.WriteString(buildDDDLayerRules())
	gates := buildQualityGates(profile)
	if gates != "" {
		mdc.WriteString("\n")
		mdc.WriteString(gates)
	}

	sections := []ConfigSection{
		NewConfigSection("AGENTS.md", agentsContent, "Cursor agents"),
		NewConfigSection(".cursor/rules/project-conventions.mdc", mdc.String(), "Cursor rules"),
	}
	if len(model.UbiquitousLanguage().Terms()) > 0 {
		sections = append(sections, NewConfigSection(
			".cursor/rules/domain-expert.mdc",
			buildCursorDomainExpertRule(model),
			"Domain terminology reference",
		))
	}
	return sections
}

// ---------------------------------------------------------------------------
// RooCodeAdapter
// ---------------------------------------------------------------------------

type rooMode struct {
	Slug           string `json:"slug"`
	Name           string `json:"name"`
	Description    string `json:"description"`
	RoleDefinition string `json:"roleDefinition,omitempty"`
}

func buildRoleDefinition(model *ddd.DomainModel) string {
	var b strings.Builder
	b.WriteString("You work within these bounded contexts:\n\n")
	for _, ctx := range model.BoundedContexts() {
		classification := "unclassified"
		if ctx.Classification() != nil {
			classification = string(*ctx.Classification())
		}
		fmt.Fprintf(&b, "- %s (%s): %s\n", ctx.Name(), classification, ctx.Responsibility())
	}
	b.WriteString("\nUse domain terminology consistently.")
	return b.String()
}

func buildEnhancedRoomodes(model *ddd.DomainModel) string {
	modes := []rooMode{
		{
			Slug:           "ddd-developer",
			Name:           "DDD Developer",
			Description:    "Follows domain-driven design conventions",
			RoleDefinition: buildRoleDefinition(model),
		},
	}

	for _, ctx := range model.BoundedContexts() {
		if ctx.Classification() == nil {
			continue
		}
		classification := string(*ctx.Classification())
		modes = append(modes, rooMode{
			Slug:           fmt.Sprintf("%s-context", strings.ToLower(ctx.Name())),
			Name:           fmt.Sprintf("%s Context", ctx.Name()),
			Description:    fmt.Sprintf("Work within %s bounded context (%s)", ctx.Name(), classification),
			RoleDefinition: fmt.Sprintf("Focus on %s. %s", ctx.Name(), ctx.Responsibility()),
		})
	}

	data := map[string]interface{}{"customModes": modes}
	roomodesJSON, _ := json.MarshalIndent(data, "", "  ")
	return string(roomodesJSON)
}

// RooCodeAdapter generates AGENTS.md, .roomodes, and .roo/rules/project-conventions.md.
type RooCodeAdapter struct{}

// NewRooCodeAdapter creates a RooCodeAdapter.
func NewRooCodeAdapter() *RooCodeAdapter { return &RooCodeAdapter{} }

// Translate implements ToolAdapter.
func (a *RooCodeAdapter) Translate(model *ddd.DomainModel, profile vo.StackProfile) []ConfigSection {
	agentsContent := buildAgentsMD(model, profile)

	roomodesJSON := buildEnhancedRoomodes(model)

	var rules strings.Builder
	rules.WriteString("# Project Conventions\n\n")
	rules.WriteString(buildDDDLayerRules())
	gates := buildQualityGates(profile)
	if gates != "" {
		rules.WriteString("\n")
		rules.WriteString(gates)
	}

	return []ConfigSection{
		NewConfigSection("AGENTS.md", agentsContent, "Roo Code agents"),
		NewConfigSection(".roomodes", roomodesJSON, "Roo Code modes"),
		NewConfigSection(".roo/rules/project-conventions.md", rules.String(), "Roo Code rules"),
	}
}

// ---------------------------------------------------------------------------
// OpenCodeAdapter
// ---------------------------------------------------------------------------

// OpenCodeAdapter generates AGENTS.md, .opencode/rules/project-conventions.md, and opencode.json.
type OpenCodeAdapter struct{}

// NewOpenCodeAdapter creates an OpenCodeAdapter.
func NewOpenCodeAdapter() *OpenCodeAdapter { return &OpenCodeAdapter{} }

// Translate implements ToolAdapter.
func (a *OpenCodeAdapter) Translate(model *ddd.DomainModel, profile vo.StackProfile) []ConfigSection {
	agentsContent := buildAgentsMD(model, profile)

	var rules strings.Builder
	rules.WriteString("# Project Conventions\n\n")
	rules.WriteString(buildDDDLayerRules())
	gates := buildQualityGates(profile)
	if gates != "" {
		rules.WriteString("\n")
		rules.WriteString(gates)
	}

	opencodeData := map[string]interface{}{
		"context": map[string]interface{}{
			"include": []string{"AGENTS.md", ".opencode/rules/*.md"},
		},
	}
	opencodeJSON, _ := json.MarshalIndent(opencodeData, "", "  ")

	return []ConfigSection{
		NewConfigSection("AGENTS.md", agentsContent, "OpenCode agents"),
		NewConfigSection(".opencode/rules/project-conventions.md", rules.String(), "OpenCode rules"),
		NewConfigSection("opencode.json", string(opencodeJSON), "OpenCode config"),
	}
}

// Compile-time interface checks.
var (
	_ ToolAdapter = (*ClaudeCodeAdapter)(nil)
	_ ToolAdapter = (*CursorAdapter)(nil)
	_ ToolAdapter = (*RooCodeAdapter)(nil)
	_ ToolAdapter = (*OpenCodeAdapter)(nil)
)
